eth/tracers/native/call_flat.go

```diff
@@ -0,0 +1,371 @@
+	tracer            *callTracer
+	config            flatCallTracerConfig
+	ctx               *tracers.Context // Holds tracer context data
+	reason            error            // Textual reason for the interruption
+	activePrecompiles []common.Address // Updated on CaptureStart based on given rules
+}
+
+type flatCallTracerConfig struct {
+	ConvertParityErrors bool `json:"convertParityErrors"` // If true, call tracer converts errors to parity format
+}
+
+// newFlatCallTracer returns a new flatCallTracer.
+func newFlatCallTracer(ctx *tracers.Context, cfg json.RawMessage) (tracers.Tracer, error) {
+	var config flatCallTracerConfig
+	if cfg != nil {
+		if err := json.Unmarshal(cfg, &config); err != nil {
+			return nil, err
+		}
+	}
+
+	tracer, err := tracers.DefaultDirectory.New("callTracer", ctx, cfg)
+	if err != nil {
+		return nil, err
+	}
+	t, ok := tracer.(*callTracer)
+	if !ok {
+		return nil, errors.New("internal error: embedded tracer has wrong type")
+	}
+
+	return &flatCallTracer{tracer: t, ctx: ctx, config: config}, nil
+}
+
+// CaptureStart implements the EVMLogger interface to initialize the tracing operation.
+func (t *flatCallTracer) CaptureStart(env *vm.EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
+	t.tracer.CaptureStart(env, from, to, create, input, gas, value)
+	// Update list of precompiles based on current block
+	rules := env.ChainConfig().Rules(env.Context.BlockNumber, env.Context.Random != nil, env.Context.Time)
+	t.activePrecompiles = vm.ActivePrecompiles(rules)
+}
+
+// CaptureEnd is called after the call finishes to finalize the tracing.
+func (t *flatCallTracer) CaptureEnd(output []byte, gasUsed uint64, err error) {
+	t.tracer.CaptureEnd(output, gasUsed, err)
+	// Parity trace considers only reports the gas used during the top call frame which doesn't include
+	// tx processing such as intrinsic gas and refunds.
+	t.tracer.callstack[0].GasUsed = gasUsed
+}
+
+// CaptureState implements the EVMLogger interface to trace a single step of VM execution.
+func (t *flatCallTracer) CaptureState(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, rData []byte, depth int, err error) {
+	t.tracer.CaptureState(pc, op, gas, cost, scope, rData, depth, err)
+}
+
+// CaptureFault implements the EVMLogger interface to trace an execution fault.
+func (t *flatCallTracer) CaptureFault(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, depth int, err error) {
+	t.tracer.CaptureFault(pc, op, gas, cost, scope, depth, err)
+}
+
+// CaptureEnter is called when EVM enters a new scope (via call, create or selfdestruct).
+func (t *flatCallTracer) CaptureEnter(typ vm.OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+	t.tracer.CaptureEnter(typ, from, to, input, gas, value)
+
+	// Child calls must have a value, even if it's zero.
+	if t.tracer.callstack[len(t.tracer.callstack)-1].Value == nil && value == nil {
```

Comment:
After fixing this for delegate calls I think the only remaining case is STatic calls. I.e. static call passes in nil for the value to CaptureEnter. I'm wondering if we can pass in 0 instead?

Reply:
Passing `0` is correct. On the other hand we don't have to match Parity that always sets a value for child calls, that said we can also remove this if and leave value to nil.

Reply:
Let's keep this for now.